packages/flutter/lib/src/rendering/sliver.dart

```diff
@@ -1742,46 +1440,599 @@ class RenderViewport2 extends RenderBox with ContainerRenderObjectMixin<RenderSl
+// /// - [RenderBox], which explains more about the Box protocol.
+// /// - [RenderSliverToBoxAdapter], which allows a [RenderBox] object to be
+// ///   placed inside a [RenderSliver] (the opposite of this class).
+// /// - [RenderShrinkWrappingViewport], a variant of [RenderViewport2] that
+// ///   shrink-wraps its contents along the main axis.
+class RenderViewport2 extends RenderViewportBase2<SliverPhysicalContainerParentData> {
+  /// Creates a viewport for [RenderSliver] objects.
+  ///
+  /// If the [center] is not specified, then the first child in the `children`
+  /// list, if any, is used.
+  ///
+  /// The [offset] must be specified. For testing purposes, consider passing a
+  /// [new ViewportOffset.zero] or [new ViewportOffset.fixed].
+  RenderViewport2({
+    AxisDirection axisDirection: AxisDirection.down,
+    @required ViewportOffset offset,
+    double anchor: 0.0,
+    List<RenderSliver> children,
+    RenderSliver center,
+  }) : _anchor = anchor,
+       _center = center,
+       super(axisDirection: axisDirection, offset: offset) {
+    assert(anchor != null);
+    assert(anchor >= 0.0 && anchor <= 1.0);
+    addAll(children);
+    if (center == null && firstChild != null)
+      _center = firstChild;
+  }
+
+  @override
+  void setupParentData(RenderObject child) {
+    if (child.parentData is! SliverPhysicalContainerParentData)
+      child.parentData = new SliverPhysicalContainerParentData();
+  }
+
+  double get anchor => _anchor;
+  double _anchor;
+  set anchor(double value) {
+    assert(value != null);
+    assert(value >= 0.0 && value <= 1.0);
+    if (value == _anchor)
+      return;
+    _anchor = value;
+    markNeedsLayout();
+  }
+
+  RenderSliver get center => _center;
+  RenderSliver _center;
+  set center(RenderSliver value) {
+    if (value == _center)
+      return;
+    _center = value;
+    markNeedsLayout();
+  }
+
+  @override
+  bool get sizedByParent => true;
+
+  @override
+  void performResize() {
+    assert(constraints.hasBoundedHeight && constraints.hasBoundedWidth);
+    size = constraints.biggest;
+    switch (axis) {
+      case Axis.vertical:
+        offset.applyViewportDimension(size.height);
+        break;
+      case Axis.horizontal:
+        offset.applyViewportDimension(size.width);
+        break;
+    }
+  }
+
+  // Out-of-band data computed during layout.
+  double _minScrollExtent;
+  double _maxScrollExtent;
+  bool _hasVisualOverflow = false;
+
+  @override
+  void performLayout() {
     if (center == null) {
       assert(firstChild == null);
-      return result;
+      _minScrollExtent = 0.0;
+      _maxScrollExtent = 0.0;
+      _hasVisualOverflow = false;
+      offset.applyContentDimensions(0.0, 0.0);
+      return;
     }
+
     assert(center.parent == this);
-    assert(firstChild != null);
+    assert(constraints.hasBoundedHeight && constraints.hasBoundedWidth);
+    double extent;
+    double crossExtent;
```

Comment:
I tend to use longer names for these (e.g., `mainAxisExtent` and `crossAxisExtent`), but I can see the argument for using these shorter names.  I'm not sure how consistent we want to be.